File: src/main/java/org/janelia/saalfeldlab/paintera/meshes/MeshInfo.java
package org.janelia.saalfeldlab.paintera.meshes;

import org.janelia.saalfeldlab.paintera.meshes.managed.MeshManagerWithSingleMesh;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.invoke.MethodHandles;

public class MeshInfo<T> {

	private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

	private final MeshManagerWithSingleMesh<T> meshManager;

	public MeshInfo(final MeshManagerWithSingleMesh<T> meshManager) {

		this.meshManager = meshManager;
	}

	public MeshSettings getMeshSettings() {

		return meshManager.getSettings();
	}

	public MeshManagerWithSingleMesh<T> meshManager() {

		return this.meshManager;
	}

	public T getKey() {

		return this.meshManager.getMeshKey();
	}
}

File: src/main/kotlin/org/janelia/saalfeldlab/paintera/meshes/MeshInfo.kt
package org.janelia.saalfeldlab.paintera.meshes

import javafx.beans.binding.ObjectExpression
import javafx.beans.property.BooleanProperty
import javafx.beans.property.ReadOnlyObjectWrapper
import javafx.util.Subscription
import org.janelia.saalfeldlab.paintera.meshes.managed.MeshManager

open class MeshInfo<T>(val key: T, open val manager: MeshManager<T>) {

	private var subscription: Subscription? = null

	val meshStateProperty: ObjectExpression<MeshGenerator.State?> by lazy {
		val property = ReadOnlyObjectWrapper<MeshGenerator.State?>()
		subscription = manager.subscribeToMeshState(key) { property.set(it) }
		property.readOnlyProperty
	}

	fun dispose() {
		subscription?.unsubscribe()
		subscription = null
	}

	override fun hashCode() = key.hashCode()

	override fun equals(other: Any?): Boolean {
		val aClass: Class<out MeshInfo<*>?> = this.javaClass
		return other != null && aClass == other.javaClass && key == aClass.cast(other)?.key
	}

	val meshSettings: MeshSettings
		get() = manager.getSettings(key)

	val isManagedProperty: BooleanProperty
		get() = manager.managedSettings.managedIndividuallyProperty(key)
}
